LibJS: Goodbye variables

davidot opened the pull request:
Implements a lot of scoping variables and more since a lot broken.
Things left to do / look at / could be done in another PR?
- [x] Fix the parser to not produce duplicated names where not allowed ~(other PR?)~ 
              Nope this is necessary for proper behavior
- [x] Classes
- [x] Switches
- [x] While loops
- [x] Refactoring
- [ ] ~Add tests to `test-js`~ Already existed
- [x] Check the other parts of the system which use LibJS
- [x] Split the commits up further? I already tried my hardest here.

Also this supersedes #10029 since that was kind of where I started and it did not really get any comments.

~I'm currently at only `-369 ✅    +205 ❌   +164 💥️`. Which is much better than the `-10000` or something I started with.~

Things which are currently broken.
- [x] Some AnnexB things (250+ regressions)
- [x] Native function calls wipe environment (100+ regression)
- [x] Switch statement scoping escapes switch in parser (??)
- [x] Fix `test-js`


Final comments:
Bytecode is probably broken.
Just 2 test262 regression to do with realms.

davidot commented:
These are the changes as I have them.
Still have to split this up into atomic commits
This has just two regression related to realm things which I didn't want to try to fix since it out of scope.

davidot commented:
I will reword the commits but it should be ready now. The code is not super clean everywhere but it is quite hard having to rebase everything with almost every LibJS commit.
I already have some plans for refactoring and using completion more but that will come later.

This has just 2 regression in test262:
```
test/language/eval-code/indirect/realm.js                            ✅ -> ❌
test/language/types/reference/get-value-prop-base-primitive-realm.js ✅ -> ❌
```
Which are related to realms.

davidot commented:
Also bytecode is probably completely broken as I mostly removed broken code but have not added in any new code.

Dexesttp commented on Userland/Libraries/LibJS/AST.cpp:
```diff
@@ -444,129 +475,274 @@ Value ForStatement::execute(Interpreter& interpreter, GlobalObject& global_objec
+                    return throw_completion(exception->value());
+            }
+        } else {
+            state.lhs_kind = ForInOfHeadState::LexicalBinding;
```
This seems to match "Step 2: If uninitializedBoundNames is not an empty List, then" in the 14.7.5.6 spec.
But I don't understand _at all_ why you're doing that after checking whether `variable_declaration.declaration_kind() == DeclarationKind::Var` instead of checking if `variable_declaration.declarations().is_empty()`, like the spec seems to require?

Edit: In general, I don't really understand what this is supposed to match in the spec... THis part kinda matches step 2, but the rest of the method doesn't seem to match anything in 14.7.5.6 (unless I really don't understand the code)...

davidot replied:
Yeah a similar comments as with the other this was a weird combination of spec which is hopefully a little more clear with the   comments now.

Dexesttp reviewed (commented):
Didn't go anywhere near reviewing everything in this chonker,  but here are a few comments about the tippy top of the page.

awesomekling commented:
This does not work correctly:
```js
var a = true;
if (a)
    switch (a) {
        case true: console.log("OK");
    }
```

It works if you do this instead:
```js
var a = true;
if (a) {
    switch (a) {
        case true: console.log("OK");
    }
}
```

So there's something that goes wrong with non-block statements.

davidot commented:
I have fixed up that LibJS was broken in some of the commits and fixed the switch issue. This now means interpreter has no scoping or executing methods anymore which was my plan anyway.
`test-js` now passes on every commit separately so I hope that is good enough. We're left with the giant commit which I hope is reviewable...

davidot commented on Userland/Libraries/LibJS/AST.cpp:
```diff
@@ -444,129 +475,274 @@ Value ForStatement::execute(Interpreter& interpreter, GlobalObject& global_objec
+        auto& for_declaration = static_cast<VariableDeclaration const&>(*expression_lhs);
+        auto& binding_pattern = for_declaration.declarations().first().target().get<NonnullRefPtr<BindingPattern>>();
+        VERIFY(lhs_kind == VarBinding || iteration_environment);
+        return interpreter.vm().binding_initialization(binding_pattern, next_value, iteration_environment, global_object);
```
No I abuse the fact that other steps above lead to either `iterator_environment == nullptr` or it has an environment. I have added some spec comments a comment to clarify this.

davidot commented on Userland/Libraries/LibJS/AST.cpp:
```diff
@@ -444,129 +475,274 @@ Value ForStatement::execute(Interpreter& interpreter, GlobalObject& global_objec
+    ThrowCompletionOr<void> execute_head(Interpreter& interpreter, GlobalObject& global_object, Value next_value) const
+    {
+        // Steps 6.g through 6.k
+        VERIFY(!next_value.is_empty());
```
I have added some spec comments and other comments to try to clarify it a bit. Unfortunately the rest of for-of/for-in loops don' really follow the spec which is why I separated out these parts. I have added a comment for that as well but I don't really want to fix that as this PR is already quite large and it follows the spec behavior pretty closely.

alimpfard commented on Userland/Libraries/LibJS/AST.h:
```diff
@@ -137,6 +139,25 @@ class ExpressionStatement final : public Statement {
+
+    template<typename CallableType>
+    IteratorOrVoidFunction(CallableType&& callable) requires(IteratorFunction<CallableType, Args...>)
+        : Function<IterationDecision(Args...)>(move(callable))
```
`callable` is a forwarding reference, use `forward()` instead of `move()`.
Also above.

davidot replied:
Yeah it was complaining about `forward` when I used that. But now it works so :shrug:. Thanks!

linusg reviewed (commented):
still need to review the big commit, here's some feedback for all the others. Message of the last commit is very misleading and makes it sound like switch is completely broken right now ("only the default clause was executed" - we have tests proving that also catch clauses work), please reword that.

linusg commented on Userland/Libraries/LibJS/AST.cpp:
```diff
@@ -283,6 +361,11 @@ Value SuperCall::execute(Interpreter& interpreter, GlobalObject& global_object)
 
     // 6. Let result be ? Construct(func, argList, newTarget).
     auto& function = new_target.as_function();
+    if (!function.has_constructor()) {
+        vm.throw_exception<TypeError>(global_object, ErrorType::NotAConstructor, "Super constructor");
+        return {};
+    }
```
We have `Value::is_constructor()`, just use it for step 5?

linusg commented on Userland/Libraries/LibJS/AST.cpp:
```diff
@@ -444,129 +527,315 @@ Value ForStatement::execute(Interpreter& interpreter, GlobalObject& global_objec
+// For the same reason we also skip step 6 and 7 of ForIn/OfHeadEvaluation as this is done by the appropriate for loop type.
+static ThrowCompletionOr<ForInOfHeadState> for_in_of_head_execute(Interpreter& interpreter, GlobalObject& global_object, Variant<NonnullRefPtr<ASTNode>, NonnullRefPtr<BindingPattern>> lhs, Expression const& rhs)
+{
+
```
Please don't add a newline after the function body start (or any scope start for that matter)

linusg commented on Userland/Libraries/LibJS/AST.cpp:
```diff
@@ -444,129 +527,315 @@ Value ForStatement::execute(Interpreter& interpreter, GlobalObject& global_objec
+                    return throw_completion(exception->value());
+            }
+        } else {
+
```
same as above, please no empty line

linusg commented on Userland/Libraries/LibJS/AST.cpp:
```diff
@@ -1017,12 +1342,27 @@ void ASTNode::dump(int indent) const
-        for (auto& variable : m_variables)
-            variable.dump(indent + 2);
+        outln("(Lexical declarations)");
+        for (auto& decl : m_lexical_declarations)
```
decl -> declaraion everywhere in this function

linusg commented on Userland/Libraries/LibJS/AST.cpp:
```diff
@@ -2666,4 +3164,181 @@ bool ImportStatement::has_bound_name(StringView name) const
+            return IterationDecision::Break;
+
+        if (restricted_global)
+            interpreter.vm().throw_exception<SyntaxError>(global_object, ErrorType::FixmeAddAnErrorStringWithMessage, "Resticted global property");
```
Resticted -> Restricted

linusg commented on Userland/Libraries/LibJS/Parser.cpp:
```diff
@@ -2834,44 +2860,70 @@ NonnullRefPtr<Statement> Parser::parse_for_statement()
+            auto& variable = declaration.declarations().first();
+            if (variable.init()) {
+                if (m_state.strict_mode || declaration.declaration_kind() != DeclarationKind::Var || !variable.target().has<NonnullRefPtr<Identifier>>())
+                    syntax_error("variable initializer not allowed in for..in/of");
```
variable -> Variable

linusg commented on Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp:
```diff
@@ -423,20 +426,252 @@ ThrowCompletionOr<Value> perform_eval(Value x, GlobalObject& caller_realm, Calle
+
+        auto* this_environment = lexical_environment;
+        while (this_environment != variable_environment) {
+
```
pointless empty line

linusg commented on Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp:
```diff
@@ -423,20 +426,252 @@ ThrowCompletionOr<Value> perform_eval(Value x, GlobalObject& caller_realm, Calle
 
-    auto& interpreter = vm.interpreter();
+    auto strict_eval = strict_caller == CallerMode::Strict;
+    if (program->is_strict_mode()) {
+        strict_eval = true;
+    }
```
no need for curly braces here

linusg commented on Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp:
```diff
@@ -423,20 +426,252 @@ ThrowCompletionOr<Value> perform_eval(Value x, GlobalObject& caller_realm, Calle
+            auto* this_environment = lexical_environment;
+
+            while (this_environment != variable_environment) {
+                if (!is<ObjectEnvironment>(*this_environment) && this_environment->has_binding(function_name)) {
+                    return IterationDecision::Continue;
+                }
```
no need for curly braces here

linusg commented on Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp:
```diff
@@ -107,86 +108,269 @@ void ECMAScriptFunctionObject::visit_edges(Visitor& visitor)
 }
 
-FunctionEnvironment* ECMAScriptFunctionObject::create_environment(FunctionObject& function_being_invoked)
+// 9.1.2.4 NewFunctionEnvironment ( F, newTarget ), https://tc39.es/ecma262/#sec-newfunctionenvironment
+FunctionEnvironment* ECMAScriptFunctionObject::new_function_environment(FunctionObject&, Object* new_target)
```
Why is this a method but still receiving the `F` argument? That doesn't seem right - `F` should be `this`.

linusg commented on Userland/Libraries/LibJS/Runtime/VM.cpp:
```diff
@@ -137,294 +137,334 @@ Symbol* VM::get_global_symbol(const String& description)
 
-void VM::assign(const Variant<NonnullRefPtr<Identifier>, NonnullRefPtr<BindingPattern>>& target, Value value, GlobalObject& global_object, bool first_assignment, Environment* specific_scope)
+// 7.3.25 CopyDataProperties ( target, source, excludedItems ), https://tc39.es/ecma262/#sec-copydataproperties
+ThrowCompletionOr<Object*> VM::copy_data_properties(Object& rest_object, Object const& source, HashTable<PropertyName, PropertyNameTraits> const& seen_names, GlobalObject& global_object)
```
Not sure why this is part of the VM, belongs to `// 7.3 Operations on Objects` in Object.cpp/h

linusg commented on Userland/Libraries/LibJS/Runtime/VM.cpp:
```diff
@@ -576,9 +609,14 @@ Value VM::construct(FunctionObject& function, FunctionObject& new_target, Option
+        return this_argument;
+
+    if (!result.is_empty() && !result.is_undefined())
+        throw_exception<TypeError>(global_object, ErrorType::DerivedConstructorReturningInvalidValue);
```
Missing return?

linusg reviewed (commented):
Fantastic work, great to see so many hacks go :^)
I probably missed some things due to the size, but since all the tests are still passing I'm not too concerned about that. The commit message would be more readable if every bullet point was indented by two spaces.

I suspect this fixes some of the open LibJS issues, will look for those and check once merged.

davidot commented:
I have rebased fixed all the mentioned problems and rewritten the switch commit message.

davidot commented:
Whoops I added spaces to the bullet points but didn't rewrap the lines.

linusg commented:
Ok let's do it! 🚀
